File: frontend/lib/api/structure.ts
import { apiPost } from "@/lib/api/client";

export type AvailableBlock = {
  category: string;
  slug: string;
  title: string;
};

export type SuggestedBlock = {
  category: string;
  slug: string;
  title: string;
  content?: Record<string, unknown> | null;
  props?: Record<string, unknown> | null;
};

export type MaxUiHint = {
  query: string;
  pattern: { id: string; name: string; sections: string[] };
  notes: string[];
  designSystem?: {
    category: string;
    severity: string;
    patternName: string;
    patternSectionsText: string;
    ctaPlacement: string;
    colorStrategy: string;
    conversionFocus: string;
    styleName: string;
    styleKeywords: string;
    styleBestFor: string;
    styleEffects: string;
    stylePerformance: string;
    styleAccessibility: string;
    colors: Record<string, string>;
    typography: Record<string, string>;
    keyEffects: string;
    antiPatterns: string;
  } | null;
  designSystemMarkdown?: string | null;
};

export type WebsiteStructurePlan = {
  blocks: SuggestedBlock[];
  notes: string[];
};

export type SuggestWebsiteStructureResponse = {
  mode: "website";
  maxUi: MaxUiHint;
  plan: WebsiteStructurePlan;
};

export async function suggestWebsiteStructure(params: {
  pageSlug: string;
  availableBlocks: AvailableBlock[];
}): Promise<SuggestWebsiteStructureResponse> {
  return apiPost<SuggestWebsiteStructureResponse>("/structure/suggest", {
    mode: "website",
    pageSlug: params.pageSlug,
    availableBlocks: params.availableBlocks,
  });
}

export type AvailableTemplate = { id: string; title: string };

export type WebappStructurePagePlan = {
  id: string;
  title: string;
  path: string;
  enabled: boolean;
  templateId?: string | null;
  objective?: string | null;
  description?: string | null;
  successCriteria?: string[];
};

export type WebappStructurePlan = {
  navigationPageIds: string[];
  pages: WebappStructurePagePlan[];
  notes: string[];
};

export type SuggestWebappStructureResponse = {
  mode: "webapp";
  maxUi: MaxUiHint;
  plan: WebappStructurePlan;
};

export async function suggestWebappStructure(params: {
  availableTemplates: AvailableTemplate[];
}): Promise<SuggestWebappStructureResponse> {
  return apiPost<SuggestWebappStructureResponse>("/structure/suggest", {
    mode: "webapp",
    availableTemplates: params.availableTemplates,
  });
}

File: backend/app/schemas/structure.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AvailableBlock(BaseModel):
    category: str
    slug: str
    title: str


class SuggestedBlock(BaseModel):
    category: str
    slug: str
    title: str
    content: dict | None = None
    props: dict | None = None


class AvailableTemplate(BaseModel):
    id: str
    title: str


class SuggestWebsiteStructureRequest(BaseModel):
    mode: Literal["website"] = "website"
    page_slug: str = Field(alias="pageSlug")
    available_blocks: list[AvailableBlock] = Field(default_factory=list, alias="availableBlocks")

    model_config = ConfigDict(populate_by_name=True)


class SuggestWebappStructureRequest(BaseModel):
    mode: Literal["webapp"] = "webapp"
    available_templates: list[AvailableTemplate] = Field(default_factory=list, alias="availableTemplates")

    model_config = ConfigDict(populate_by_name=True)


class WebsiteStructurePlan(BaseModel):
    blocks: list[SuggestedBlock] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class WebappStructurePagePlan(BaseModel):
    id: str
    title: str
    path: str
    enabled: bool = True
    template_id: str | None = Field(default=None, alias="templateId")
    objective: str | None = None
    description: str | None = None
    success_criteria: list[str] = Field(default_factory=list, alias="successCriteria")

    model_config = ConfigDict(populate_by_name=True)


class WebappStructurePlan(BaseModel):
    navigation_page_ids: list[str] = Field(default_factory=list, alias="navigationPageIds")
    pages: list[WebappStructurePagePlan] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)

    model_config = ConfigDict(populate_by_name=True)


class MaxUiPattern(BaseModel):
    id: str
    name: str
    sections: list[str] = Field(default_factory=list)


class MaxUiDesignSystem(BaseModel):
    category: str = "General"
    severity: str = "MEDIUM"
    pattern_name: str = Field(default="", alias="patternName")
    pattern_sections_text: str = Field(default="", alias="patternSectionsText")
    cta_placement: str = Field(default="", alias="ctaPlacement")
    color_strategy: str = Field(default="", alias="colorStrategy")
    conversion_focus: str = Field(default="", alias="conversionFocus")

    style_name: str = Field(default="", alias="styleName")
    style_keywords: str = Field(default="", alias="styleKeywords")
    style_best_for: str = Field(default="", alias="styleBestFor")
    style_effects: str = Field(default="", alias="styleEffects")
    style_performance: str = Field(default="", alias="stylePerformance")
    style_accessibility: str = Field(default="", alias="styleAccessibility")

    colors: dict[str, str] = Field(default_factory=dict)
    typography: dict[str, str] = Field(default_factory=dict)

    key_effects: str = Field(default="", alias="keyEffects")
    anti_patterns: str = Field(default="", alias="antiPatterns")

    model_config = ConfigDict(populate_by_name=True)


class MaxUiHint(BaseModel):
    query: str
    pattern: MaxUiPattern
    notes: list[str] = Field(default_factory=list)
    design_system: MaxUiDesignSystem | None = Field(default=None, alias="designSystem")
    design_system_markdown: str | None = Field(default=None, alias="designSystemMarkdown")

    model_config = ConfigDict(populate_by_name=True)


class SuggestWebsiteStructureResponse(BaseModel):
    mode: Literal["website"] = "website"
    max_ui: MaxUiHint = Field(alias="maxUi")
    plan: WebsiteStructurePlan

    model_config = ConfigDict(populate_by_name=True)


class SuggestWebappStructureResponse(BaseModel):
    mode: Literal["webapp"] = "webapp"
    max_ui: MaxUiHint = Field(alias="maxUi")
    plan: WebappStructurePlan

    model_config = ConfigDict(populate_by_name=True)
